Moving placeholder variable to EditableFormField

3Dgoo opened the pull request:
Moving placeholder to the base form field class and controlling it's appearance through a config variable. 

This addresses issue #577

robbieaverill commented:
Chuck this against the `1` branch if you can.

3Dgoo commented:
Is that the `4.3` branch?

robbieaverill commented on lang/eo.yml:
```diff
@@ -83,6 +83,7 @@ eo:
     NOTBLANK: 'Ne vaka'
     NOTVALUE: 'Ne valoro'
     OPTIONS: Agordoj
+    PLACEHOLDER: Lokokupilo
```
I'm not 100%, but I think that removing old translations from non English translations files is fine because Transifex will only merge in new values, but to add the new values they'll need to be added in Transifex - @dhensby do you know more about this?

dhensby replied:
That's right

robbieaverill commented:
@3Dgoo sorry, it's the `4` branch - my head was in another module. I've just added a comment re: the translations being shuffled around. Will see what @dhensby says about it.

This PR looks great to me otherwise though! Thanks very much!

3Dgoo commented:
Happy to contribute. Thanks for the help and guidance. 

I'll put through a new pull request with the changes against the 4 branch once we hear back on what to do with the translations.

robbieaverill commented:
@3Dgoo you can change this PR's target branch via the "Edit" button on the RHS of the PR title, no need to open another one. Will require a rebase though. Thanks!

3Dgoo commented:
Rebased and changed  the target branch. 

It was good to learn how to do that properly.

robbieaverill commented:
Thanks. It's probably OK to leave the new translations where they are for now until Transifex overwrites them later.

tractorcow commented:
One concern I have about this change is that any existing placeholder value is lost. Is there no way we could migrate the value from the subclass table to the base? Otherwise user experience could be quite negative. Perhaps a build task? It could just be a bunch of simple SQL queries for the draft / live tables for each type.

robbieaverill commented:
I think that's a pretty valid point, while it doesn't break the PHP API it does require a data migration. Would you like writing a build task for this @3dgoo? We can add it to the changelogs if they don't exist already

tractorcow commented:
You could build it into https://github.com/silverstripe/silverstripe-userforms/blob/master/code/tasks/UserFormsUpgradeService.php as another step.

3Dgoo commented:
I'll give the changes a test and add some migration code to the dev task.

3Dgoo commented:
In `UserFormsUpgradeService` there is a `migrateCustomSettings` function:
https://github.com/silverstripe/silverstripe-userforms/blob/4/code/tasks/UserFormsUpgradeService.php#L114

This calls `migrateSettings` for each field. Would it be OK to put the place holder migration code in the `migrateSettings` for `EditableTextField`, `EditableEmailField` and `EditableNumericField`?

robbieaverill commented:
@3Dgoo I reckon so

tractorcow commented:
Thanks @3Dgoo we appreciate your hard work. :)

3Dgoo commented:
Thanks. Happy to help :) 

Is it cool to use `DB::query` or should I really be using `SQLSelect` and `SQLUpdate ` instead?

3Dgoo commented:
I have added some migration code using `DB::query`. I have given this a test and it works for me. 
Have a look and let me know what you think and if you want any changes.

3Dgoo commented:
I see I have some tests to fix. I'll get on that.

3Dgoo commented:
I'm not super proud of how many nested if statements are in that placeholder migration task, but it works. If you guys want that cleaned up I'm open to suggestions.

tractorcow commented:
> I'm not super proud of how many nested if statements are in that placeholder migration task, but it works. If you guys want that cleaned up I'm open to suggestions.


The trick to this is to rewrite as a new method, and use early returns instead.

```
function methodOne() {
	$this->methodTwo();
	parent::methodOne();
}

function methodTwo() {
	if (!$cond) {
		return;
	}
	if (!$cond2) {
		return;
	}
	// do a thing
}
```

robbieaverill commented:
Is it possible to move the migration logic into the build task rather than the form field classes?

3Dgoo commented:
@tractorcow - I have improved migration using a separate method with `return` calls. Thanks for the help on that.

3Dgoo commented:
@robbieaverill - I can do that if you think that's what we should do. 

I can move the three `migrateSettings` functions to one `migrateSettings` function in the `UserFormsUpgradeService` class. I think we should call the function from within the `migrateCustomSettings` function. Does that sound good?

3Dgoo commented:
@robbieaverill - Done. I have moved the three migrateSettings functions to one migrateSettings function in the UserFormsUpgradeService class.

tractorcow commented on code/tasks/UserFormsUpgradeService.php:
```diff
@@ -224,4 +230,66 @@ public function getQuiet()
+            return;
+        }
+        // Update draft Placeholder value
+        $query = "UPDATE `EditableFormField` SET `Placeholder` = '$draftPlaceholder' WHERE `ID` = '$field->ID'";
```
Can you please SQL escape this? there's a SQL injection vulnerability here.

If possible, use DB::prepared_query, then you can use placeholders instead.

```php
DB::prepared_query(
	"UPDATE \"EditableFormField\" SET \"Placeholder\" = ? WHERE "ID" = ?",
	[ $draftPlaceholder, $field->ID]
);
```

Please use double quotes instead of back ticks if you can help it. ticks are mysql-only.

tractorcow commented on code/tasks/UserFormsUpgradeService.php:
```diff
@@ -224,4 +230,66 @@ public function getQuiet()
+            return;
+        }
+        // Check if draft table exists
+        $query = "SHOW TABLES LIKE '$tableName'";
```
also `DB::get_schema()->hasTable($tableName)` for shorthand.

also `DB::get_schema()->hasField($tableName, $fieldName)` to check both at the same time :)

3Dgoo commented:
Thanks @tractorcow. I have fixed up the sql injection issue, replaced the back ticks and improved the table and column checking code. Thanks for the tips.

Let me know if there is anything else that can be improved or fixed..

3Dgoo commented:
Awesome. Thanks very much Damian and Robbie.

robbieaverill commented:
Thanks for your hard work!
